bundles/org.openhab.binding.gardena/src/main/java/org/openhab/binding/gardena/internal/GardenaSmartImpl.java

```diff
@@ -12,562 +12,384 @@
-
-    private static final String DEVICE_CATEGORY_MOWER = "mower";
-    private static final String DEVICE_CATEGORY_GATEWAY = "gateway";
-
-    private static final String DEFAULT_MOWER_DURATION = "180";
-
-    private static final String URL = "https://smart.gardena.com";
-    private static final String URL_LOGIN = URL + "/v1/auth/token";
-    private static final String URL_LOCATIONS = URL + "/v1/locations/?user_id=";
-    private static final String URL_DEVICES = URL + "/v1/devices/?locationId=";
-    private static final String URL_COMMAND = URL + "/v1/devices/%s/abilities/%s/command?locationId=%s";
-    private static final String URL_PROPERTY = URL + "/v1/devices/%s/abilities/%s/properties/%s?locationId=%s";
-    private static final String URL_SETTING = URL + "/v1/devices/%s/settings/%s?locationId=%s";
-
-    private Gson gson = new GsonBuilder().registerTypeAdapter(Date.class, new DateDeserializer())
-            .registerTypeAdapter(PropertyValue.class, new PropertyValueDeserializer()).create();
-    private HttpClient httpClient;
+    private Gson gson = new GsonBuilder().registerTypeAdapter(DataItem.class, new DataItemDeserializer()).create();
 
-    private String mowerDuration = DEFAULT_MOWER_DURATION;
-    private Session session;
-    private GardenaConfig config;
-    private String id;
+    private static final String URL_API_HUSQUARNA = "https://api.authentication.husqvarnagroup.dev/v1";
+    private static final String URL_API_GARDENA = "https://api.smart.gardena.dev/v1";
+    private static final String URL_API_TOKEN = URL_API_HUSQUARNA + "/oauth2/token";
+    private static final String URL_API_WEBSOCKET = URL_API_GARDENA + "/websocket";
+    private static final String URL_API_LOCATIONS = URL_API_GARDENA + "/locations";
+    private static final String URL_API_COMMAND = URL_API_GARDENA + "/command";
 
+    private String id;
+    private GardenaConfig config;
     private ScheduledExecutorService scheduler;
-    private ScheduledFuture<?> refreshThreadFuture;
-    private RefreshDevicesThread refreshDevicesThread = new RefreshDevicesThread();
 
+    private Map<String, Device> allDevicesById = new HashMap<>();
+    private LocationsResponse locationsResponse;
     private GardenaSmartEventListener eventListener;
 
-    private Map<String, Device> allDevicesById = new HashMap<>();
-    private Set<Location> allLocations = new HashSet<>();
+    private HttpClient httpClient;
+    private List<GardenaSmartWebSocket> webSockets = new ArrayList<>();
+    private @Nullable PostOAuth2Response token;
+    private boolean initialized = false;
+    private WebSocketFactory webSocketFactory;
 
-    @Override
-    public void init(String id, GardenaConfig config, GardenaSmartEventListener eventListener,
-            ScheduledExecutorService scheduler) throws GardenaException {
+    private Set<Device> devicesToNotify = Collections.synchronizedSet(new HashSet<>());
```

Comment:
I'm not sure how safe the synchronized behavior is when doing iteration. It think it would be much safe here to use one of the concurrent implementations instead.
```suggestion
    private Set<Device> devicesToNotify = ConcurrentHashMap.newKeySet();
```